fix config, add main (iss. #10, #11, #13)

takposha opened the pull request:
Rolling in multiple issues as they all tie into how the config works mostly.
Added the main function, as well as a config Class that handles all validation of config variables, defaults, and error catching. This also sets up the groundwork to handle custom queries in #6, which will be a different PR. I think the scope of #12 is also changed here as the Config class will be checking for type errors, but I will get that looked at before saying that is addressed. 

Resolves #10
Resolves #11 
Resolves #13

lsloan commented on mpr-data-access.ipynb:
```diff
@@ -0,0 +1,156 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "mysql+pymysql://mwrite_peer_review_research_ro:j5L%nKVKayHCdzYL@tl-prod-mysql.aws.vdc.it.umich.edu:3306/mwrite_peer_review\n"
```
The commit of this new file has exposed the credentials (password) of the MPR DB to the _public_ repo on GH.  We need to reset the password of this DB ASAP.  I'll open a ticket to do this.

takposha replied:
I'm really sorry about this. That Jupyter NB was not even used after the first day or so because of the risk of exposing credentials. I need to be more careful about what VSCode stages for syncing.

lsloan replied:
Some git clients try to be extremely helpful with committing as much as possible.  It can be time-consuming, but it's worth it to double check everything before committing or pushing it to the online repo.

lsloan replied:
My advice WRT this kind of thing:

1. Any output in a Jupyter notebook can be saved as part of the notebook.
2. Check the outputs before saving copies of the notebook anywhere.  Remove compromising info.
3. Some Jupyter implementations let you remove some or all output cells.  Some let you specify that the output for specific cells should never be saved.
4. If you print a password for debugging purposes, once you're certain it's working, remove the print statement ASAP.

takposha replied:
Noted. I'll be more aware of this moving forward.

lsloan commented on mpr-research-data.py:
```diff
@@ -161,34 +122,156 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
+
+            if not self.dbParams[credPart]:
+                logging.error(
+                    f'Did not find .env variable for M-Write Peer Review production DB key: {credPart}.')
```
In production, these variables won't be set by a file called `.env`.  They will just be variables set in the OpenShift environment.  (Maybe from a file, but from within the environment, it won't seem to be.)  It's probably better to say "configuration item" rather than ".env variable".

takposha replied:
Noted. I will change that.

lsloan replied:
✅

lsloan reviewed (changes requested):
Pretty good.  I made comments on some things I think could be improved.  I may give a few more comments soon.

takposha commented on testQuery.sql:
```diff
@@ -0,0 +1,41 @@
+-- https://docs.google.com/document/d/1FEDjYng1FilD64xx_XHct1zff_Ds9T9hsswFw-p2_J4/edit#bookmark=id.3znysh7
+SELECT
+  canvas_courses.name AS 'Course',
+  canvas_courses.id AS 'CourseID',
+  prompts.title AS 'Prompt',
+  prompts.id AS 'PromptID',
+  authors.sortable_name AS 'Author',
+  authors.id AS 'AuthorID',
+  reviewers.sortable_name AS 'Reviewer',
+  reviewers.id AS 'ReviewerID',
+  criteria.description AS 'Criterion',
+  criteria.id AS 'CriterionID',
+  peer_review_comments.comment AS 'Comment',
+  peer_review_comments.id AS 'CommentID',
+  peer_review_comments.commented_at_utc AS 'CommentTimeUTC'
+FROM
+  peer_review_comments
+LEFT JOIN peer_reviews ON
+  peer_review_comments.peer_review_id = peer_reviews.id
+LEFT JOIN canvas_students AS reviewers ON
+  peer_reviews.student_id = reviewers.id
+LEFT JOIN canvas_submissions ON
+  peer_reviews.submission_id = canvas_submissions.id
+LEFT JOIN canvas_assignments AS prompts ON
+  canvas_submissions.assignment_id = prompts.id
+LEFT JOIN canvas_courses ON
+  prompts.course_id = canvas_courses.id
+LEFT JOIN rubrics ON
+  prompts.id = rubrics.reviewed_assignment_id
+LEFT JOIN canvas_students AS authors ON
+  canvas_submissions.author_id = authors.id
+LEFT JOIN criteria ON
+  peer_review_comments.criterion_id = criteria.id
+WHERE
+  canvas_courses.id in ({})
+ORDER BY
+  prompts.id,
+  authors.sortable_name,
+  reviewers.sortable_name,
+  peer_review_comments.criterion_id
```
It's a residual query I'm using for testing some things on eventual configurable queries and how the code reads them in. I will remove it (and put it in .gitignore) for future commits if I keep it for purely testing uses.

lsloan replied:
✅

takposha commented on mpr-research-data.py:
```diff
@@ -161,34 +122,156 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
+        try:
+            self.gcpParams = json.loads(str(os.getenv('GCP_KEY')))
+            return True
+        except Exception as e:
```
My assumption here for this approach is that we are trapping the error we think is most likely/expect to occur?

takposha replied:
As a follow-up, would an error in something such as failing to connect to the Database or GCP, where it could be more than just one expected error, do we try something else?

lsloan replied:
Apologies for the late response…

lsloan replied:
Apologies for the late response…

Yes, trap the exception that is most likely to occur.  In this case, as explained by the `json.loads()` documentation.  In testing, if we find that an additional exception occurs, we can add another `except` clause to handle it.  It's better to handle each exception individually than to catch every possible one with the generic `Exception`.

For connecting to the DB or GCP resources, catch the basic exceptions that their documentation says is likely to occur.  If we find other exceptions occur during testing, those could be handled with more `except` clauses.  If by "try something else" you mean trying to reconnect to the resource, I think that's not necessary.  It would be helpful to log messages about it, though.

pep8speaks commented:
Hello @takposha! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`mpr-research-data.py`](https://github.com/tl-its-umich-edu/mpr-research-data/blob/2454baba23b261b209088e5b41b44da4ad35953a/mpr-research-data.py):

> [Line 24:1](https://github.com/tl-its-umich-edu/mpr-research-data/blob/2454baba23b261b209088e5b41b44da4ad35953a/mpr-research-data.py#L24): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

lsloan commented on mpr-research-data.py:
```diff
@@ -7,65 +7,34 @@
+    raiseLogTypeError = None
 
+    try:
+        logLevel = str(os.environ.get('LOG_LEVEL', 'DEBUG')).upper()
```
We should use a default level of `INFO`, I think.  That shouldn't be too verbose for production.  For development, we can set the level to `DEBUG` in our own `.env` files.

lsloan replied:
✅

lsloan commented on mpr-research-data.py:
```diff
@@ -7,65 +7,34 @@
+    if logLevel not in ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'FATAL', 'CRITCAL', 'EXCEPTION']:
+        logWarning = True
 
-def queryRetriever(queryName, queryModifier=False,
```
IMNSHO, it appears GH isn't doing a good job with diffs.   Showing lines of the other function which had been more than 30 lines later mixed in with lines of this doesn't make the diff very easy to read, does it?

(Just an observation.  There's nothing we can do about it.)

lsloan commented on mpr-research-data.py:
```diff
@@ -7,65 +7,34 @@
+        logWarning = True
+        raiseLogTypeError = e
+    
+    if logLevel not in ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'FATAL', 'CRITCAL', 'EXCEPTION']:
+        logWarning = True
```
A couple of thoughts about this log level validation…

1. As I mentioned in comments for #15, maybe don't validate it.  Try using it and fall back to a default level (`INFO`) if an exception is thrown about an invalid value.
2. If you think this validation is important and should be done here, there's a good programmatic way to get the valid level names:
   ```suggestion
       if logLevel not in [logging.getLevelName(10 * i) for i in range(1, 6)]:
           logWarning = True
   ```

   That gives the five basic log levels:

   ```pycon
   >>> [logging.getLevelName(10 * l) for l in range(1, 6)]
   ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']
   ```
   
   I think `EXCEPTION` isn't a log level.  If it's something new, please point out the documentation to me.  The level `FATAL` is an alias for `CRITICAL`, so it's unnecessary.

takposha replied:
I'm just going to remove validation to simplify things, I like the idea to keep it pythonic.

lsloan replied:
✅

lsloan reviewed (changes requested):
Logging config comments.

lsloan commented on mpr-research-data.py:
```diff
@@ -161,34 +143,167 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
+        try:
+            self.numberOfMonths = str(
+                os.getenv('NUMBER_OF_MONTHS', self.numberOfMonths))
+            if not self.numberOfMonths.isnumeric():
```
Maybe even better…

```python
        try:
            self.numberOfMonths = int(
                os.getenv('NUMBER_OF_MONTHS', self.numberOfMonths))
        except ValueError as e:
```

Try to convert with `int()` and catch `ValueError` exceptions.

lsloan replied:
✅

takposha commented:
Changelog: 

- Changed wording of error messages to better reflect error causes
- Added lower bound check to NUMBER_OF_MONTHS
- Added int check to DB_PORT
- Simplified logger setup
- Added catch for empty course list - exit if no courses are found

lsloan commented on mpr-research-data.py:
```diff
@@ -144,51 +107,228 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
+            self.numberOfMonths = int(
+                os.getenv('NUMBER_OF_MONTHS', self.numberOfMonths))
+
+            if self.numberOfMonths < 1:
+                logging.error(
+                    f'Config parameter NUMBER_OF_MONTHS must be >= 1.')
+                envImportSuccess = False
```
This check is very good.  I had noticed it myself shortly before I saw you added it.

However, I just tested this check by setting months to 0 and I got this output…

```
mpr-research-data | Starting mpr-research-data.py...
mpr-research-data | 2022-05-13T19:27:46+00:00
mpr-research-data | 2022-05-13T19:27:48+0000 ERROR [mpr-research-data.py:250] - Config parameter NUMBER_OF_MONTHS must be >= 1.
mpr-research-data | 2022-05-13T19:27:48+0000 INFO [mpr-research-data.py:156] - Loading in SQL file - courseQuery.sql
mpr-research-data | 2022-05-13T19:27:48+0000 INFO [mpr-research-data.py:156] - Loading in SQL file - retrieveQuery.sql
mpr-research-data | Exiting due to configuration parameter import problems.
mpr-research-data | 🏁 - Python Script Execution Complete.
mpr-research-data | 2022-05-13T19:27:48+00:00
mpr-research-data exited with code 0
```

It seems to me that the application should've exited after the error about the months and before the queries are loaded.

takposha replied:
This is partly intentional. All config variables are checked even if one has failed. This way, if multiple config values have issues, they can all be logged, rather than doing it over multiple runs. I can change that if that is not preferred.

lsloan replied:
Oh, I see.  Yes, that's a pretty good approach.  You don't need to change in that case.

lsloan commented on mpr-research-data.py:
```diff
@@ -144,51 +107,228 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
 
-# RETRIEVE KEYS
-# --------------------------------------------------------------------------
+class Config:
```
I think some of the config item validation code is repetitive.  I wonder whether some of the validation is similar enough that a generalized version of it could be used to apply to multiple config items.  It may even reduce the amount of code.

I'm thinking (not thoroughly tested code here) something like:

```python
def configFetch(name, default=None, casting=None, validation=None):
  value = os.getenv(name, default)
  if (casting is not None):
    value = casting(value)
  if (validation is not None and not validation(value)):
    raise ValueError(f'Validation error for config item "{name}" value "{value}".')
  return value

# A practical application, I think
self.numberOfMonths = configFetch('NUMBER_OF_MONTHS', self.numberOfMonths, 
  int, lambda x: x > 0)

# Maybe contrived, but an example of how flexible this could be.
# Demonstrates the default and validation may be skipped, and the casting may be complex
self.gcpParams = configFetch('GCP_KEY', casting=lambda x: json.loads(str(x)))
```

Maybe my suggestion could even be a reinvention of the wheel.  I'd suggest looking around on [PyPI](http://pypi.python.org/) to see whether some modules already exist for fetching and validating values from environment variables.

takposha replied:
I have never used a lambda function like that before! That's extremely powerful. I will implement this in the config setup.

lsloan replied:
I see from your commit (a416e7b) that it seems like for the config handling, you ended up cutting the number of lines of code almost in half.

lsloan reviewed (changes requested):
Validation of config items can be tricky.  We've done quite a bit with it in the CCM and MyLA projects, but they're based on TypeScript and Django, respectively, and I think they can't be reused here.

takposha commented:
Changelog:
- Added function for config variable checking
- Added better System Exit messages
- 

To-do:
- Possibly add better Error Messages when a Validation error happens, as now it does not specify. A dictionary should be enough for this.
- Add an immediate stop to config setup if invalid config is found (if decided to do so)

lsloan commented on mpr-research-data.py:
```diff
@@ -145,140 +149,72 @@ def set(self, name, value):
+                errorMsg = f'Casting error for config item "{name}" value "{value}".'
+                logging.error(errorMsg)
+                return None
+                #raise TypeError(errorMsg)
```
So, for this commented `raise` line and the one a few lines down, what's your thinking?  Do you feel more comfortable handling the `None` return values than catching exceptions?  If you think that's the way to go, you should definitely delete the commented lines.

takposha replied:
I like to handle Nones better as I do check if all config values have been set up correctly, even if one already failed, as a way to find all errors on file in one go. I will remove the commented lines.

lsloan replied:
OK.  `None` it is, then.  I'll watch for the comment removal.

lsloan replied:
✅

lsloan reviewed (changes requested):
Returning `None` vs. throwing exception.

lsloan commented on mpr-research-data.py:
```diff
@@ -144,51 +107,228 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
+        except TypeError as e:
+            logging.warning(f'Error Message: {e}')
+            logging.warning(
+                'Incorrect type for configuration parameter for log level provided. Defaulting to INFO level.')
```
✅

lsloan commented on mpr-research-data.py:
```diff
@@ -144,51 +107,164 @@ def sliceAndPushToGCP(courseDF, retrieveDF, client, targetBucketName=targetBucke
 
-logging.info('All steps complete.')
+        try:
+            self.logLevel = str(os.environ.get('LOG_LEVEL', self.logLevel)).upper()
```
Here, `os.environ.get()` is used.  In `configFetch()`, `os.getenv()` is used (line 153).  There are minor differences between the two functions.  One suggestion I saw was that `os.getenv()` **_shouldn't_** be used only because its corresponding `os.putenv()` is broken.  So it's better to use the functions in the `os.environ` module to avoid mistakes.

Look it up and decide for yourself.  Either way, we should use the same function on both lines for getting the values of the environment variables.

takposha replied:
Will fix that to `os.environ.get()` everywhere.

lsloan replied:
Sounds good to me!  ✅

lsloan commented on mpr-research-data.py:
```diff
@@ -8,66 +8,14 @@
-    format="%(asctime)s %(levelname)-8s [%(filename)s:%(lineno)d] - %(message)s",
+    format="%(asctime)s %(levelname)-4s [%(filename)s:%(lineno)d] - %(message)s",
     datefmt="%Y-%m-%dT%H:%M:%S%z",
     level=logging.INFO
```
✅

takposha commented:
Changelog:
Added better error handling. SQL Alchemy wraps almost all errors into one call, so that should cover most cases.
Redid GCP connection initialization - This was primarily because it was not actually checking anything until a file was attempted to be written. So I check for a bucket existence while setting up as a way to ensure that works first. 
Minor edits and removal of commented lines as per feedback in PR #14

lsloan reviewed (approved):
I like the additional early checks for the DB and GCP connections.

lsloan commented:
I notice that the initial unsigned commit has come to haunt this PR, too, just as it did the previous one.  I didn't expect this.  I wouldn't think that commit would appear as part of this PR.  Maybe it's something to do with how the `das-v2` branch was created.  We can review this, if you'd like.
